### PR TITLE
[CoreML EP] Add Flatten Op and LRN Op support

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
### Description

As title.

CoreML Spec for reference: 

https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#flattento2dlayerparams

https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#lrnlayerparams

### Motivation and Context

Fill CoreML Clipchamp usage gaps.

### CODE REVIEW

File: onnxruntime/core/providers/coreml/builders/impl/flatten_op_builder.cc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "Flatten does not support empty input shape";
+    return false;
+  }
+
+  const auto input_rank = input_shape.size();
+  if (input_rank < 3) {
+    LOGS(logger, VERBOSE) << "Flatten only supports input rank greater than equal to 3, input rank is"
+                          << input_rank;
+    return false;
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/flatten_op_builder.cc:75]: Is this a requirement for flatten? didn't notice it in the spec and the example show it working with input that has a rank of 1.

Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/builders/impl/flatten_op_builder.cc:75]: good catch. the requirement comes from FlattenLayerParams - was originally looking at that one in CoreML. For FlattenTo2D, it does not require the input rank >= 3.

**Review by skottmckay — APPROVED** [review]

File: onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "LRN does not support empty input shape";
+    return false;
+  }
+
+  const auto input_rank = input_shape.size();
+  if (input_rank < 3) {
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc:76]: from the [CoreML LRN spec](https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#lrnlayerparams):

"For ranks greater than 3, the leading dimensions, starting from 0 to -4 (inclusive), are all treated as batch."

I don't think [ONNX LRN](https://github.com/onnx/onnx/blob/main/docs/Operators.md#LRN) also does this. should we limit it to supporting input rank 3 or 4?

Comment (YUNQIUGUO) [reply onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc:76]: okay I see. will update.

Comment (skottmckay) [reply onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc]: Good catch.

File: onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "LRN does not support empty input shape";
+    return false;
+  }
+
+  // Note: For higher ranks ( > 3), CoreML LRN treats all leading dimentions as a match,
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc:75]:
```suggestion
  // Note: For higher ranks ( > 3), CoreML LRN treats all leading dimensions as a match,
```

**Review by edgchen1 — DISMISSED** [review]

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**